patches/system-appearance-change-emacs-27.patch

```diff
@@ -0,0 +1,234 @@
+-                            appearanceNamed: NSAppearanceNameVibrantDark];
+-      FRAME_NS_APPEARANCE (f) = ns_appearance_vibrant_dark;
++#if MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
++#ifndef NSAppKitVersionNumber10_14
++#define NSAppKitVersionNumber10_14 1671
++#endif
++      if (NSAppKitVersionNumber >= NSAppKitVersionNumber10_14)
++        {
++          window.appearance = [NSAppearance
++                                  appearanceNamed: NSAppearanceNameDarkAqua];
++          FRAME_NS_APPEARANCE (f) = ns_appearance_dark_aqua;
++        }
++      else
++#endif /* MAC_OS_X_VERSION_MAX_ALLOWED >= 101400 */
++        {
++          window.appearance = [NSAppearance
++                                  appearanceNamed: NSAppearanceNameVibrantDark];
++          FRAME_NS_APPEARANCE (f) = ns_appearance_vibrant_dark;
++        }
+     }
+-  else
++  else if (EQ (new_value, Qlight))
+     {
+       window.appearance = [NSAppearance
+                             appearanceNamed: NSAppearanceNameAqua];
+       FRAME_NS_APPEARANCE (f) = ns_appearance_aqua;
+     }
++  else
++    {
++      // Reset window appearance to track the system appearance.
++      window.appearance = nil;
++    }
+ #endif /* MAC_OS_X_VERSION_MAX_ALLOWED >= 101000 */
+ }
+ 
+@@ -7477,12 +7496,27 @@ - (instancetype) initFrameFromEmacs: (struct frame *)f
+ #if defined (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101000
+ #ifndef NSAppKitVersionNumber10_10
+ #define NSAppKitVersionNumber10_10 1343
++#endif
++#if MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
++#ifndef NSAppKitVersionNumber10_14
++#define NSAppKitVersionNumber10_14 1671
+ #endif
+ 
+-  if (NSAppKitVersionNumber >= NSAppKitVersionNumber10_10
+-      && FRAME_NS_APPEARANCE (f) != ns_appearance_aqua)
+-    win.appearance = [NSAppearance
+-                          appearanceNamed: NSAppearanceNameVibrantDark];
++  if (NSAppKitVersionNumber < NSAppKitVersionNumber10_14)
++#endif /* MAC_OS_X_VERSION_MAX_ALLOWED >= 101400 */
++    if (NSAppKitVersionNumber >= NSAppKitVersionNumber10_10)
++      {
++        if (FRAME_NS_APPEARANCE(f) != ns_appearance_aqua)
++          {
++            win.appearance =
++                [NSAppearance appearanceNamed:NSAppearanceNameVibrantDark];
++          }
++        else
++          {
++            win.appearance =
++                [NSAppearance appearanceNamed:NSAppearanceNameAqua];
++          }
++    }
+ #endif
+ 
+ #if defined (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101000
+@@ -8223,6 +8257,41 @@ - (instancetype)toggleToolbar: (id)sender
+   return self;
+ }
+ 
++#if defined (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
++#ifndef NSAppKitVersionNumber10_14
++#define NSAppKitVersionNumber10_14 1671
++#endif
++- (void)viewDidChangeEffectiveAppearance
++{
++  NSTRACE ("[EmacsView viewDidChangeEffectiveAppearance:]");
++
++  if (NSAppKitVersionNumber < NSAppKitVersionNumber10_14)
++    return;
++
++  // The `ns-appearance' frame parameter overrides the system appearance;
```

Comment:
Do I understand it correctly, that by default `ns-appearance` is not set, which means that with this patch, Emacs appearance will change between dark and light depending on the system configuration?

Reply:
Yes; As long as the frame parameter `ns-appearance` is not explicitly set to `'dark` or `'light`, the appearance of the frame will follow the global system appearance, and the functions added to the new hook variable will be run.